src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,445 @@
+                break;
+            case 12000:
+                madUnits = MAD_UNITS_12000_HZ;
+                break;
+            case 16000:
+                madUnits = MAD_UNITS_16000_HZ;
+                break;
+            case 22050:
+                madUnits = MAD_UNITS_22050_HZ;
+                break;
+            case 24000:
+                madUnits = MAD_UNITS_24000_HZ;
+                break;
+            case 32000:
+                madUnits = MAD_UNITS_32000_HZ;
+                break;
+            case 44100:
+                madUnits = MAD_UNITS_44100_HZ;
+                break;
+            case 48000:
+                madUnits = MAD_UNITS_48000_HZ;
+                break;
+            default:
+                qWarning() << "Invalid sample rate:" << m_file.fileName()
+                        << madSampleRate;
+                return ERR; // abort
+            }
+            setFrameRate(madSampleRate);
+        } else {
+            // check for consistent frame/sample rate
+            if ((0 < madSampleRate) && (getFrameRate() != madSampleRate)) {
+                qWarning() << "Differing sample rate in some headers:"
+                        << m_file.fileName()
+                        << getFrameRate() << "<>" << madSampleRate;
+                return ERR; // abort
+            }
+        }
+
+        mad_timer_add(&madDuration, madHeader.duration);
+        sumBitrate += madHeader.bitrate;
+        mad_header_finish(&madHeader);
+
+        // Add next frame to list of frames
+        addSeekFrame(madDuration, madUnits, m_madStream.this_frame);
+    }
+
+    if (1 < m_seekFrameList.size()) {
+        setFrameCount(mad_timer_count(madDuration, madUnits));
+        m_avgSeekFrameCount = getFrameCount() / (m_seekFrameList.size() - 1);
+        int avgBitrate = sumBitrate / (m_seekFrameList.size() - 1);
+        setBitrate(avgBitrate / 1000);
+    } else {
+        // This is not a working MP3 file.
+        qWarning() << "SSMP3: This is not a working MP3 file:"
+                << m_file.fileName();
+        return ERR; // abort
+    }
+
+    restartDecoding(m_seekFrameList.front());
+
+    return OK;
+}
+
+void AudioSourceMp3::restartDecoding(const SeekFrameType& seekFrame) {
+    // Reset the MAD decoder completely. Otherwise
+    // audible artifacts and glitches occur when
+    // seeking through the stream no matter how
+    // many MP3 frames are prefetched!
+    mad_synth_finish(&m_madSynth);
+    mad_frame_finish(&m_madFrame);
+    mad_stream_finish(&m_madStream);
+    mad_stream_init(&m_madStream);
+    mad_stream_options(&m_madStream, MAD_OPTION_IGNORECRC);
+    mad_frame_init(&m_madFrame);
+    mad_synth_init(&m_madSynth);
+
+    // Reset stream position
+    m_curFrameIndex = seekFrame.frameIndex;
+
+    // Discard decoded output
+    m_madSynthCount = 0;
+
+    // Fill input buffer
+    mad_stream_buffer(&m_madStream, seekFrame.pInputData,
+            m_fileSize - (seekFrame.pInputData - m_pFileData));
+
```

Comment:
here we need a mad_header_decode(&m_madFrame->header, &m_madStream) call as well.